src/server.js

```diff
@@ -0,0 +1,299 @@
+/* @flow */
+/* global ErrorCat */
+'use strict'
+
+const assign = require('101/assign')
+const clone = require('101/clone')
+const defaults = require('101/defaults')
+const errorCat = require('error-cat')
+const Immutable = require('immutable')
+const isFunction = require('101/is-function')
+const isObject = require('101/is-object')
+const pick = require('101/pick')
+const Promise = require('bluebird')
+
+const logger = require('./logger')
+const RabbitMQ = require('./rabbitmq')
+const Worker = require('./worker')
+
+/**
+ * Ponos server class. Given a queue adapter the worker server will
+ * connect to RabbitMQ, subscribe to the given queues, and begin spawning
+ * workers for incoming jobs.
+ *
+ * The only required option is `opts.queues` which should be a non-empty flat
+ * list of strings. The server uses this list to subscribe to only queues you
+ * have provided.
+ *
+ * @author Bryan Kendall
+ * @author Ryan Sandor Richards
+ * @param {Object} opts Options for the server.
+ * @param {ErrorCat} [opts.errorCat] An error cat instance to use for the
+ *   server.
+ * @param {Object<String, Function>} [opts.events] Mapping of event (fanout)
+ *   exchanges which to subscribe and handlers.
+ * @param {bunyan} [opts.log] A bunyan logger to use for the server.
+ * @param {Object} [opts.rabbitmq] RabbitMQ connection options.
+ * @param {String} [opts.rabbitmq.hostname] Hostname for RabbitMQ.
+ * @param {Number} [opts.rabbitmq.port] Port for RabbitMQ.
+ * @param {String} [opts.rabbitmq.username] Username for RabbitMQ.
+ * @param {String} [opts.rabbitmq.password] Username for Password.
+ * @param {Object<String, Function>} [opts.tasks] Mapping of queues to subscribe
+ *   directly with handlers.
+ */
+class Server {
+  _events: Map<string, Function>;
+  _opts: Object;
+  _rabbitmq: any;
+  _tasks: Map<string, Function>;
+  _workerOptions: Object;
+  errorCat: ErrorCat;
+  log: Object;
+
+  constructor (opts: Object) {
+    this._opts = assign({}, opts)
+    this.log = this._opts.log || logger.child({ module: 'ponos:server' })
+    this._workerOptions = {}
+
+    this._tasks = new Immutable.Map()
+    if (this._opts.tasks) {
+      this.setAllTasks(this._opts.tasks)
+    }
+    this._events = new Immutable.Map()
+    if (this._opts.events) {
+      this.setAllEvents(this._opts.events)
+    }
+
+    this.errorCat = this._opts.errorCat || errorCat
+
+    this._rabbitmq = new RabbitMQ(this._opts.rabbitmq || {})
+  }
+
+  /**
+   * Start consuming from the subscribed queues. This is called by `.start`.
+   * This can be called after the server has been started to start consuming
+   * from additional queues.
+   *
+   * @return {Promise} Promise resolved when consuming has started.
+   */
+  consume (): Promise {
```

Comment:
Are there instances where you would want to call this from outside the class? If not, should we make this method private?